### PR TITLE
feat!: capture group bindings from metavariable-regex

### PR DESCRIPTION
**Pull request by brandonspark**
## What:
This PR makes it so that capture groups used in `metavariable-regex` regexes can now persist after the `metavariable-regex`.

## Why:
This is a big wanted feature, and will make our rule writing experience better.

## How:
Added logic which will factorize out the `metavariable-regex` stuff with `Xpattern_matcher_regexp`, and produce matches which will be passed up, in a similar way to `Metavariable_pattern`.

## Test plan:
`make test`

Closes PA-3011 PRO-120

PR checklist:

- [X] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [X] Tests included or PR comment includes a reproducible test plan
- [X] Documentation is up-to-date
- [X] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [X] Change has no security implications (otherwise, ping security team)

If you're unsure about any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

### CODE REVIEW

**Comment by aryx** [comment]
Can this be done independently of the remove the $1, $2 ? I see the stacked-diff tag.

**Comment by brandonspark** [comment]
@aryx we will incur a test failure if we do, because of a rule which is currently failing because of implicit metavariable unification for the reasons i mentioned in #8515

**Comment by aryx** [comment]
Is this still a stacked diff?
Can you rebase on your other numeric_do_not_unify_mvar fix?

**Comment by brandonspark** [comment]
@aryx this actually already is rebased on top of the other PR (see that the merge base is different). It's ready for review now, I think.

File: src/engine/Eval_generic.ml
Comment (aryx) [inline src/engine/Eval_generic.ml]: Please use Fpath.t instead.

Comment (aryx) [reply src/engine/Eval_generic.ml]: and also can you add a comment what this file refers to? Why do we need it?

File: src/engine/Metavariable_regex.mli
Comment (aryx) [inline src/engine/Metavariable_regex.mli]: Please use Fpath.t in new code.

File: src/engine/Metavariable_regex.mli
Comment (aryx) [inline src/engine/Metavariable_regex.mli]: Is this comment still valid?
The first parameter (not argument)  is an environment, not a function ...

**Review by aryx — CHANGES_REQUESTED** [review]

**Comment by aryx** [comment]
failing in CI

**Comment by brandonspark** [comment]
@aryx was just the snapshots, should be good now

**Comment by brandonspark** [comment]
nvm there's more hold up

**Comment by brandonspark** [comment]
@aryx I investigated and found that these errors are due to a lack of ported functionality to `osemgrep scan`. In particular, we are not properly deduplicating findings as the Python CLI does (see https://github.com/returntocorp/semgrep/blob/28820ed668db4eea3bb98144294ccd12418c1118/src/osemgrep/reporting/Cli_json_output.ml#L609).

This seems like a fair amount of code which is unrelated to the current feature, which simply surfaces an instance where this lack of deduplication is biting us. Would it be fair to simply remove this test from the list of `osempass` expected tests, and move on with merging (especially given that this is quite high priority)?

**Comment by aryx** [comment]
ok remove the osempass, and maybe add a comment next to it with a TODO and link to this PR that introduced the regression, so another person can investigate it.

**Review by akuhlens — APPROVED** [review]
LGTM, @aryx feel free to follow up with @brandonspark with any further changes.

**Comment by akuhlens** [comment]
Apparently I can't approve this if @aryx hasn't. I tried based off what I read as tacit approval. Sorry if I stepped on any toes.

File: cli/tests/e2e/snapshots/test_check/test_metavariable_propagation_regex/results.json
Comment (aryx) [inline cli/tests/e2e/snapshots/test_check/test_metavariable_propagation_regex/results.json:25]: I thought we were actually considering the numeric metavar differently, and set unify_numeric_metavar to false.

File: libs/commons/Regexp_engine.mli
Comment (aryx) [inline libs/commons/Regexp_engine.mli:26]: you could extend pcre_compile with a ?flags instead of defining a separate function

File: src/engine/Eval_generic.ml
Comment (aryx) [inline src/engine/Eval_generic.ml:61]: typo which

**Review by aryx — APPROVED** [review]

**Pull request merged**
